backends/ubpf/ubpfControl.cpp

```diff
@@ -0,0 +1,772 @@
+            emitDeclaration(builder, a);
+        }
+
+        codeGen->setBuilder(builder);
+        builder->emitIndent();
+        controlBlock->container->body->apply(*codeGen);
+
+        for (size_t i = 0; i < codeGen->registersLookups.size(); i++) {
+            builder->newline();
+            builder->blockEnd(true);
+        }
+    }
+
+    const IR::Statement *UBPFControl::findFirstStatementWhereVariableIsUsedAsPointer(
+            const IR::Statement *statement,
+            const IR::Declaration_Variable *vd) {
+        if (statement->is<IR::IfStatement>()) {
+            auto ifStat = statement->to<IR::IfStatement>();
+            auto result = findFirstStatementWhereVariableIsUsedAsPointer(
+                    ifStat->ifTrue, vd);
+            if (result == nullptr) {
+                return findFirstStatementWhereVariableIsUsedAsPointer(
+                        ifStat->ifFalse, vd);
+            } else {
+                return result;
+            }
+        }
+        if (statement->is<IR::BlockStatement>()) {
+            auto block = statement->to<IR::BlockStatement>();
+            for (auto cmpnt : block->components) {
+                if (cmpnt->is<IR::Statement>()) {
+                    auto result = findFirstStatementWhereVariableIsUsedAsPointer(
+                            cmpnt->to<IR::Statement>(), vd);
+                    if (result != nullptr) {
+                        return result;
+                    }
+                }
+            }
+        }
+        if (statement->is<IR::AssignmentStatement>() &&
+            variableIsUsedAsPointer(vd,
+                                    statement->to<IR::AssignmentStatement>())) {
+            return statement;
+        }
+
+        return nullptr;
+    }
+
+    void UBPFControl::emitDeclaration(EBPF::CodeBuilder *builder,
+                                      const IR::Declaration *decl) {
+        if (decl->is<IR::Declaration_Variable>()) {
+            auto vd = decl->to<IR::Declaration_Variable>();
+            auto etype = UBPFTypeFactory::instance->create(vd->type);
+            builder->emitIndent();
+
+            bool pointerType = determineIfVariableIsPointer(vd, etype);
+
+            if (pointerType) {
+                codeGen->pointerVariables.push_back(vd->name.name);
+            }
+
+            etype->declare(builder, vd->name, pointerType);
+            builder->endOfStatement(true);
+            BUG_CHECK(vd->initializer == nullptr,
+                      "%1%: declarations with initializers not supported",
+                      decl);
+            return;
+        } else if (decl->is<IR::P4Table>() ||
+                   decl->is<IR::P4Action>() ||
+                   decl->is<IR::Declaration_Instance>()) {
+            return;
+        }
+        BUG("%1%: not yet handled", decl);
+    }
+
+    bool UBPFControl::determineIfVariableIsPointer(const IR::Declaration_Variable *vd, const EBPF::EBPFType *etype) {
```

Comment:
Wouldn't it be cleaner to have a separate Inspector pass run before your code generation and build a map of all variables that should be used as pointers?

Reply:
That's good idea. I've just refactored the code to build vector of pointer variables before emitting them.